fix(iceberg): ensure ChunkInfoRowNumbers in position delete compactio…

kdt523 opened the pull request:
fix(iceberg): add RowNumbersTransform to ensure ChunkInfoRowNumbers in position delete pipeline

 Changelog category:
- Bug Fix (user-visible misbehavior in an official stable release)

Changelog entry:
Fixed an issue where Iceberg table compaction with position deletes would fail due to missing row number information. Added a RowNumbersTransform to ensure proper row number tracking during compaction operations.

Documentation entry for user-facing changes:

- Documentation is not required for this bugfix

Technical Details 
Changes made:
- Add new RowNumbersTransform to ensure row numbers are properly set
- Add transform to pipeline before position delete operations
- Improve error messages to help diagnose missing ChunkInfoRowNumbers

This fix ensures that position delete operations in Iceberg tables work correctly by making sure row number information is always available in the compaction pipeline. Error messages are now more descriptive to help users understand and debug any potential issues.

The change follows existing ClickHouse patterns for transform implementation and pipeline modification, maintaining code consistency while fixing the functionality.
Refs: #88123 
contributing for hactoberfest 2025

kdt523 committed:
fix(iceberg): ensure ChunkInfoRowNumbers in position delete compaction pipeline

alexey-milovidov reviewed (changes requested):
Missing tests.

scanhex12 commented on src/Storages/ObjectStorage/DataLakes/Iceberg/PipelineUtils.h:
```diff
@@ -0,0 +1,15 @@
+    {
+        return std::make_shared<RowNumbersTransform>(header.cloneEmpty());
+    });
+}
```
please add new line at the end of file

alexey-milovidov replied:
Resolved differently: `PipelineUtils.h` (and `RowNumbersTransform.h`) are removed in the latest push -- the fix moved into `Compaction.cpp::writeDataFiles` instead, since that is where compaction actually reads data files. See the PR comment for details.

scanhex12 commented on src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp:
```diff
@@ -884,6 +885,12 @@ void IcebergMetadata::addDeleteTransformers(
     if (!iceberg_object_info->position_deletes_objects.empty())
     {
+        // Add RowNumbersTransform before position delete transform
+        builder.addSimpleTransform([](const Block & header)
```
Could you please explain why do we need this transform? As I understood it just calculates `current_row_num` and save it in the private field and that's all.

kdt523 replied:
Position deletes match rows by absolute position, and PositionDeleteTransform reads those positions from ChunkInfoRowNumbers on each Chunk. Parquet readers populate this metadata, but the ORC path via icebergLocal doesn’t—so without it the transform fails (“ChunkInfoRowNumbers does not exist”) or can’t apply deletes correctly.

RowNumbersTransform computes and attaches that metadata (using the running current_row_num offset) so position deletes can be applied reliably. It’s only added when position deletes are present, and if the metadata is already there it’s effectively a no‑op. Verified by the new test tests/integration/test_storage_iceberg_no_spark/test_position_deletes_row_numbers.py (ORC + position deletes + OPTIMIZE).

alexey-milovidov replied:
Good catch -- this `RowNumbersTransform` usage in `IcebergMetadata.cpp` (the `SELECT` read pipeline) turned out not to fix the reported crash on current `master`, since compaction reads data files directly in `Compaction.cpp::writeDataFiles` rather than going through that pipeline. Reworked to attach `ChunkInfoRowNumbers` there instead, where a running row offset over a single-stream, physical-order read is provably correct (no pipeline-level parallel/pre-filtering concern). Details in the PR comment.

scanhex12 reviewed (commented):
Thanks for contribution, but I didn't understand the problem that was fixed. Could you please add a test?

kdt523 committed:
tests(iceberg): add ORC+position-deletes optimize regression test; fix newline at EOF in PipelineUtils.h

alexey-milovidov commented:
Hi @kdt523! Thanks for the contribution. The PR description is missing the Changelog category and entry sections from the [PR template](https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md). Could you please add them? Thank you!

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into fix/iceberg-row-numbers

# Conflicts:
#	src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp
#	src/Storages/ObjectStorage/DataLakes/Iceberg/PositionDeleteTransform.cpp

alexey-milovidov committed:
Fix Iceberg compaction crash on ORC data files with position deletes

`OPTIMIZE` on an Iceberg table converts equality deletes into position
deletes and applies them while rewriting each data file in `writeDataFiles`
(`Compaction.cpp`). The position-delete transform maps physical row positions
to a deletion bitmap using `ChunkInfoRowNumbers`. Parquet and row-based readers
attach this info to every chunk, but block formats such as ORC do not, so
compacting a table containing an ORC data file failed with the exception
`Logical error: 'ChunkInfoRowNumbers does not exist'`.

The original change in this pull request added a transform on the read
(`SELECT`) pipeline in `IcebergMetadata::addDeleteTransformers`, but compaction
does not go through that pipeline - it reads each data file directly and calls
the transform itself - so it did not fix the reported issue on current `master`.

Fix it in `writeDataFiles`: compaction reads each data file with a single
stream in physical order, so the running row offset is the correct absolute
position. Attach `ChunkInfoRowNumbers` to each chunk when the reader did not
provide it. The misplaced `RowNumbersTransform.h` and the unused `PipelineUtils.h`
are removed, and the diagnostics that report a missing `ChunkInfoRowNumbers` now
explain that the data reader must attach row numbers.

Closes: https://github.com/ClickHouse/ClickHouse/issues/88123

alexey-milovidov committed:
Merge remote-tracking branch 'origin/master' into fix/iceberg-row-numbers

alexey-milovidov commented:
Thanks for the report and the regression test, @kdt523! I reworked the fix on top of current `master`, since the branch had drifted significantly and the original approach no longer addressed the crash.

The original `RowNumbersTransform` was inserted into `IcebergMetadata::addDeleteTransformers`, i.e. the `SELECT` read pipeline. On current `master`, compaction (`OPTIMIZE`) does not go through that pipeline — `Compaction.cpp::writeDataFiles` reads each data file directly and applies the delete transform itself — so that transform never ran during compaction and the original `Logical error: 'ChunkInfoRowNumbers does not exist'` still reproduced.

I moved the fix into `writeDataFiles`: it reads each data file with a single stream in physical order, so a running row offset is the correct absolute row position, and `ChunkInfoRowNumbers` is now attached there when the reader (e.g. `ORC`) did not provide it. This also resolves @scanhex12's correctness concern about the transform's naive sequential counter not matching physical positions under parallel/pre-filtered reads — that scenario doesn't apply here since compaction's per-file read is always a single stream. The stray `RowNumbersTransform.h`/`PipelineUtils.h` files are removed, and I kept and adjusted your integration test, which reproduces the original bug and passes with the fix (verified locally).

Note there is a parallel fix for the same root issue in #110107, which takes a different approach (skip the delete transform entirely for data files that have no position deletes attached, since only those can be non-Parquet). Maintainers may prefer to land one or the other — worth a look if this one is picked up for review.

Merged current `master` in to resolve the conflicts and keep this buildable; build and the integration test pass locally.

alexey-milovidov commented:
Fixed in https://github.com/ClickHouse/ClickHouse/pull/110107
